### PR TITLE
feat(datasource/crate): fetch crate metadata from crates.io

### LINKED ISSUE #3486: Add sourceUrl/homepage lookup from crates API for cargo datasource
**What would you like Renovate to be able to do?**
Add sourceUrl/homepage lookup from crates API with 24 hour caching.
Note that crates.io has a crawler policy forbidding more than 1 request per second,
so the solution should respect this restriction.
**Additional context**
This is related to #3442.
See [this comment](https://github.com/renovatebot/renovate/pull/3474#issuecomment-478915448).

### PR DESCRIPTION
**Pull request by Turbo87**
## Changes

This PR adds changelogs to Rust crate update PRs by downloading crate metadata from crates.io. It is using a fast path on the crates.io server to avoid unnecessary database requests (see code comment) and work around the regular request per second limitation. Since the metadata rarely changes, it will be cached for 24 hours to avoid unnecessary API requests.

Bildschirmfoto 2022-04-21 um 00 23 21

Example PR from this branch: https://github.com/Turbo87/renovate-crates-test/pull/3

## Context

Closes https://github.com/renovatebot/renovate/issues/3486

Note that I have seen https://github.com/renovatebot/renovate/pull/14841, but since the discussion there was not concluded yet, I've implemented it only for the crates.io datasource for now.

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/crate/index.spec.ts [hunk 1]
```diff
@@ -58,6 +60,13 @@ function setupErrorGitMock(): { mockClone: jest.Mock<any, any> } {
   return { mockClone };
 }
 
+function mockCratesApiCallFor(crateName: string, response?: httpMock.Body) {
+  httpMock
+    .scope(API_BASE_URL)
+    .get(`/crates/${crateName}?include=`)
+    .reply(response ? 200 : 404, response ?? '');
```
Comment (viceice) [inline lib/modules/datasource/crate/index.spec.ts:67]:
```suggestion
    .reply(response ? 200 : 404, response);
```
should work too?

Comment (Turbo87) [reply lib/modules/datasource/crate/index.spec.ts]: yep, indeed 👍

**Review by rarkins — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Comment by Turbo87** [comment]
I assume it'll take a while before this is deployed to the GitHub app?

**Comment by rarkins** [comment]
Yeah, unfortunately a while - maybe next week. There's some stuff which is holding me back on an earlier version

**Comment by lfrancke** [comment]
Thank you both for the work and super quick turnaround!

**Pull request merged**